Add a contributing guide (#323)
holoviz/holoviz: doc/contributing.md

# Contributing Guide

This guide includes general guidelines on how to contribute to HoloViz. Read the [Operating Guide](TODO) if you are interested to learn how an Open-Source project like HoloViz operates.

## Why contribute?

There are many reasons why you would contribute to the HoloViz project, including:

> * You appreciate the HoloViz project, its mission and values, and would like to help

Reviewer: should we add a link to 'mission and values'... maybe the `background` or `about` page?

### Code

Contributing code includes:

* fixing an existing bug: the recommended practice is to add one or more unit test(s) that would have failed before the bug fix was implemented.
* adding a new feature: again, the recommended practice is to add unit test(s) to test the new feature's functionality.
* improving performance: while slow code isn't necessarily a bug, users always appreciate faster code. Performance improvements should be demonstrated by some benchmark comparing the performance before and after the changes.
* adding tests: test coverage can always be improved. Adding tests is a good way for beginners to contribute to a project and familiarize themselves with its code base and workflows.
* refactoring: refactoring can be a good way to improve a code base.

### Developer Instructions

```{note}
The HoloViz projects usually provide specific instructions to set up your environment to be ready to contribute documentation or code to the project. The instructions provided hereafter are meant to be general enough to apply to each project. Refer to the documentation of the project you are working on for more details.
```

> Before making any contribution, the first step is usually to open a [GitHub Issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-issues) that documents what you are trying to fix/improve. Once an issue is opened, a discussion can take place with the maintainers to see if what is suggested is relevant, which can orientate the action to take.

Reviewer: ```suggestion
Before making a substantial contribution, the first step is usually to open a [GitHub Issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-issues) that documents what you are trying to fix/improve. Once an issue is opened, a discussion can take place with the maintainers to see if what is suggested is relevant, which can orientate the action to take.
```
This should be consistent with the statement later on:
> An issue should be created before submitting a pull request (PR) unless the scope of the planned PR is minimal, such as fixing a typo.

Author: Isn't `substantial` too strong here?

Reviewer: fair enough, how about this: 

```suggestion
Before making anything beyond a minimal contribution (e.g. fixing a typo), the first step is usually to open a [GitHub Issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-issues) that documents what you are trying to fix/improve. Once an issue is opened, a discussion can take place with the maintainers to see if what is suggested is relevant, which can orientate the action to take.
```

Reviewer 2: I'd also suggest pointing folks to a blog post that describes what a good issue should look like, e.g. https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

As an external contributor, you do not have permission to submit code directly to any of the repositories. Github offers you to fork the repository, which will create a copy of the repository, on which you will have the right to work freely. Follow [these instructions](https://docs.github.com/en/get-started/quickstart/fork-a-repo) to find out how to fork and clone a repository.

The HoloViz developers rely on [Conda](https://conda.io/docs/intro.html) to manage their virtual environments. We **recommend** you install Conda too to have an experience as close as possible to those of the core developers. To install Conda on any platform, you can [install Miniconda](https://docs.conda.io/en/latest/miniconda.html). 

The HoloViz developers rely on [pyctdev](https://github.com/pyviz-dev/pyctdev), a custom developer tool that facilitates maintaining all of the holoviz projects. pyctdev is an extension of the task running tool [doit](https://pydoit.org/). We **recommend** that you install `pyctdev`:

```bash
conda install -c pyviz/label/dev "pyctdev>0.5.0"
```

> Once `pyctdev` is available and you are in the cloned panel repository, you can create a new virtual environment with the `env_create` *doit* command with the appropriate channels:

Reviewer: ```suggestion
Once `pyctdev` is available and you are in the cloned (e.g. panel) repository, you can create a new virtual environment with the `env_create` *doit* command with the appropriate channels:
```

```bash

> doit env_create -c channel1 -c channel2 --name=dev-env --python=3.x

Reviewer: maybe add an example ~ ('For example: `doit env_create -c pyviz -c conda-forge --python=3.8 --name=my_dev_env`')

```

Don't forget to activate this environment:

```bash
conda activate dev-env
```

To perform an editable install of the project you are working on, including all the dependencies required to run the full unit test suite, run the `develop_install` with the appropriate channels and options:

```bash

> doit develop_install -c channel1 -c channel2 -o tests -o ... -o ...

Reviewer: maybe add an example ~ ('For example: `doit -c pyviz -c conda-forge develop_install -o tests -o examples`')

```

At this step, you should have your environment set up, being able to run tests and a clone of the source repository. The next step is to create a branch and start making changes. Before committing these changes, make sure the tests still pass by running them. The HoloViz source codes are stringent concerning styling (e.g., they're not using [black](https://github.com/psf/black)). Try to write code that follows the style of the surrounding code. Once you are done with your changes, you can commit them. It is good practice to break down big changes into smaller chunks, and each chunk has its own commit with a message that summarizes it. Now you can push the branch to your clone and create a *pull request* from your clone to the original repository. This [Github Gist](https://gist.github.com/Chaser324/ce0505fbed06b947d962) describes these GitHub steps in more detail.

The *pull request* you make should reference the *issue* you are attempting to close (i.e. `Fixes #issuenumber`) and include a description of the changes you made. Changes affecting the visual properties should include screenshots or GIFs.

> Your *pull request* should now be reviewed by one of the project's maintainers. This may take a while given how busy the maintainers are, so try to be patient :) Once the review is done you may be required to make some changes. You may also be asked to *resolve conflicts* if the changes you made appear to conflict with changes made to the source code after you submitted the *pull request*.

Reviewer: ```suggestion
Your *pull request* should now be reviewed by one of the project's maintainers. This may take a while given how busy the maintainers are, so try to be patient :). Once the review is done you may be required to make some changes. You may also be asked to *resolve conflicts* if the changes you made appear to conflict with changes made to the source code after you submitted the *pull request*.
```

> Your *pull request* should now be reviewed by one of the project's maintainers. This may take a while given how busy the maintainers are, so try to be patient :) Once the review is done you may be required to make some changes. You may also be asked to *resolve conflicts* if the changes you made appear to conflict with changes made to the source code after you submitted the *pull request*.

Reviewer: ```suggestion
Your *pull request* should now be reviewed by one of the project's maintainers. This may take a while given how busy the maintainers are, so try to be patient :). We will evaluate the pull request to ensure that it is within the scope of the project (again, it's a good idea to create an issue first to outline your intent and give us an opportunity to weigh in before you spend the effort creating a pull request). If we decide to proceed, we will conduct a review, and once the review is done you may be required to make some changes. You may also be asked to *resolve conflicts* if the changes you made appear to conflict with changes made to the source code after you submitted the *pull request*.
```

> Your *pull request* should now be reviewed by one of the project's maintainers. This may take a while given how busy the maintainers are, so try to be patient :) Once the review is done you may be required to make some changes. You may also be asked to *resolve conflicts* if the changes you made appear to conflict with changes made to the source code after you submitted the *pull request*.

Author: ```suggestion
Your *pull request* should now be reviewed by one of the project's maintainers. This may take a while given how busy the maintainers are, so try to be patient :). Your pull request will be evaluated to ensure that it is within the scope of the project (again, it's a good idea to create an issue first to outline your intent and give the maintainers an opportunity to weigh in before you spend the effort creating a pull request). If it's decided to proceed, the review will be conducted, and once the review is done you may be required to make some changes. You may also be asked to *resolve conflicts* if the changes you made appear to conflict with changes made to the source code after you submitted the *pull request*.
```

*PyViz*, or [PyViz.org](https://pyviz.org/), is now a community project of its own. It is a fully open guide to all Python visualization tools. The HoloViz group maintains the [pyviz](https://github.com/pyviz) organization, but only for historical reasons, as it is meant to be shared by the whole Python visualization community.

### GitHub/Git

#### Organisations and repositories

HoloViz consists of several Python projects. All these projects are version-controlled with [Git](https://git-scm.com/) and hosted on GitHub. Having all the projects hosted on GitHub means that HoloViz can rely on all the excellent features Github freely provides to open-source projects, including easy collaboration through *Issues and Pull Requests (PR)*, continuous integration (CI) with *GitHub Actions (GHA)* and documentation hosting with *GitHub Pages*.

```{warning}

> It is crucial to keep in mind that nothing lasts forever and that, at some point, HoloViz may be forced to rely on other platforms or services. The more HoloViz relies on GitHub, the more difficult any transition to another system is.

Reviewer: why is this a warning to users/readers? Isn't it an internal warning for the core developers?

Author: The Operating Guide is meant to be read by contributors (or anyone interested in learning how such a project is maintained, for instance someone from another such project). In theory any contributor can become a core developer, the more we share the better imo.

Reviewer: I really like the point, I just don't know if this is the correct place or tag for it. This is already a very extensive document, and I think `warning` tags should imply imminent danger and justify extra attention. If you want to retain the text as highlighted box, maybe change it to something less alarming, like a `note`? It's not a big deal, feel free to resolve it either way

Author: I agree it'd better be a *note*.

```

The HoloViz group owns a few GitHub organizations:

* [holoviz](https://github.com/holoviz/) is the main one where you are likely to contribute. It hosts the core packages maintained by the group.
* [pyviz-dev](https://github.com/pyviz-dev/) hosts two main types of repositories:

>     * Packages that support maintaining the core HoloViz packages, including, for instance, `nbsite`, `nbsmoke`, `pyctdev`, `pyct` and `autover`. These support packages were developed when no alternative, or satisfying alternative, was available at the time the group needed them. Interestingly a few of these packages could have become pretty popular had they been more promoted. However, it was never a goal for the group that instead focused on improving data visualization with Python rather than improving the Python tooling.

Reviewer: ```suggestion
    * Packages that support maintaining the core HoloViz packages, including, for instance, `nbsite`, `nbsmoke`, `pyctdev`, `pyct` and `autover`. These support packages were developed when no alternative, or satisfying alternative, was available at the time the group needed them.
```
This comes off a bit weird, like 'we could have done a thing, but we didn't'. I don't think it's necessary information for readers.

Author: This is something I've heard from others in the group that's why I shared it. I agree it comes off a bit weird, so now you've also heard it I'm fine removing it!

    * Repositories that are only used to host *dev* builds of the core packages websites, i.e., no actual work is done on these repositories. They just get updated automatically in a CI process. 

> * [pyviz-topics](https://github.com/pyviz-topics/) hosts repositories that demonstrate concrete usage of the HoloViz tools.

Reviewer: I'm assuming there's probably too much friction to change the name of the `pyviz-dev` org to `holoviz-dev`, but what about changing `pyviz-topics` to `holoviz-topics` to make it consistent with its description?

Reviewer: maybe we should also combine `holoviz-demos` and `pyviz-topics` as one org for holoviz uses curated by the holoviz team? I understand this is not the place to discuss these things, but just noting a thought to bring up later

Author: I'm not sure about the state of `holoviz-demos` to be honest and where it's really used. As for renaming the orgs, there's friction for sure but it's not impossible.

* A Sphinx `NotebookDirective` allows inserting an evaluated notebook in a document. It has a few useful parameters like `offset` that takes a number that will be the number of top cells not rendered. 
* It can build a gallery from an organized collection of Notebooks.

Building a site with `nbsite` is usually a two-step process:

1. `nbsite generate-rst ...` looks for notebooks in the `examples` folder and generates their corresponding *reStructuredText* files. For instance, if the notebook `examples/user_guide/Data.ipynb` is found, then the corresponding file `doc/user_guide/Data.rst` is created and includes the `NotebookDirective` that points to the notebook file to insert it in this document. A similar process applies to the notebooks found in a gallery.
2. `nbsite build ...` executes the notebooks and builds the website.

> After these steps, you should find a `builtdocs` folder in the repository that contains the static site built by nbsite/sphinx.

Reviewer: ```suggestion
After these steps, you should find a `builtdocs` folder in the repository that contains the static site built by nbsite/sphinx. When we build the public docs website, the `builtdocs` goes to the github pages branch, as can be seen in the `.github` folder workflow.
```

Author: I try not to use `we` that much in the document, to be as inclusive as possible.

> After these steps, you should find a `builtdocs` folder in the repository that contains the static site built by nbsite/sphinx.

Author: ```suggestion
After these steps, you should find a `builtdocs` folder in the repository that contains the static site built by nbsite/sphinx. When the websites are built in the continuous integration system, the content of the `builtdocs` folder is pushed to a `gh-pages` branch. The details of this process can be found in the `docs.yaml` Github Action workflow of each project, located in the `.github/workflows` folder.
```

* *Jupyter Notebooks* (.ipynb): Notebooks are saved in the `examples` folder and are meant to be executed when the documentation is built, which means that the system that builds the documentation must have all the dependencies and data required to run them. Notebooks **must be cleared before being committed** to a repository. TODO: Share somewhere the `jq` command to clear the notebooks. Thanks to [MyST-NB](https://myst-nb.readthedocs.io/en/latest/), on which `nbsite` depends, *MyST Markdown* is correctly handled in notebooks.
* *reStructuredText* (.rst): Original file format supported by Sphinx and in which the Python documentation is written. They should be gradually replaced by *MyST Markdown* files.
* *Markdown* (.md):  [MyST Markdown](https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html#syntax-core) is a Markdown extension focused on scientific and technical documentation authoring. It is easier to write and read compared to *reStructuredText*, which should be favored.

#### Theme

The HoloViz sites rely on the [PyData Sphinx Theme](https://pydata-sphinx-theme.readthedocs.io/) for their theme.

#### Analytics

> Currently, none of the websites gather any analytics via *Google Analytics*. This used to be the case but was reverted because of the GDPR rules (TODO really?).

Reviewer: This does not appear to be a true statement yet

Author: Yep correct. I think we first need to decide if Google Analytics is actually okay with respect to the GDPR rules. We could also assess whether plausible.io would be just a better alternative. There's some information here on plausible.io: https://pydata-sphinx-theme.readthedocs.io/en/latest/user_guide/analytics.html. Maybe one of us should reach at the people who have put up this recommendation to see why exactly they recommend plausible.io over Google Analytics.

Then whatever we decide, we need to implement it. You're the one who gets the chance to set the priority of this action, i.e. if you really need that data to orient how to improve the docs or not. I've personally never used this kind of service so I don't really know how much we'd get from it. I expect we would learn interesting things though.

* *Example tests*: the examples tests and the notebooks tests, in which the notebooks found in the `/examples` folder are all executed. Note that their output is not compared with any reference. Instead, the tests only fail if an error is raised while running the notebooks. The projects rely on pytest and either [nbsmoke](https://github.com/pyviz-dev/nbsmoke) or [nbval](https://nbval.readthedocs.io). The `doit test_examples` command is usually the one that will run these tests.
* *UI tests*: some projects may rely on [Playwright](https://playwright.dev/python/) and [pytest-playwright](https://github.com/microsoft/playwright-pytest) to run tests that check that things get displayed as expected in a browser and those interactions between the client and the backend work as expected. The `doit test_ui` command is usually the one that will run these tests.

```{note}

> We would like not to have to maintain nbsmoke anymore and instead rely on nbqa and nbval.

Reviewer: what is the purpose of including this note? maybe instead this should be an internal discussion point or even an issue that gets filed.

Author: Indeed there should be a meta issue on nbsmoke's repo or in this repo that records that we'd like to get rid of nbsmoke entirely. I think the note is still useful, as the document needs to introduce `nbsmoke` as it is currently in use, but also need to warn contributors that they should try not to extend its usage.

Reviewer: Sounds good. Presumably, potential contributors to nbsmoke are more likely to read the nbsmoke readme, so maybe it should go there as well (or instead) with additional context.

* A new release is a good opportunity to check that **there are no alarming warnings** emitted while the tests suite runs. Missing a deprecation warning could mean that you would have to make a new release soon after this one!

#### Detailed process

Development releases can have different goals. Sometimes they are only meant to be pre-releases made right before an actual release to make sure everything is alright. Sometimes they are made to be shared with stakeholders (e.g., a specific contributor, a customer, a dependent project, etc.). Depending on your situation, you might decide to pause the release process in the procedure detailed below between two development releases, waiting for feedback.

1. Before making a proper release, you should start by making a development release. This is the first one that would be an *alpha* release. Bump the version too, e.g., `v1.9.6a1` and commit the new tag.
1. After pushing the new tag, you can monitor the *Packages* and *Documentation* workflows. If one fails, you will have to fix that as that would be a release blocker.
1. If the *alpha* release succeeded, it is now time to check a few things:

>     * A new version of the *dev* site should have been built. You should spot-check it, trying to find errors that could have occurred while the notebooks ran, for instance.

Reviewer: ```suggestion
    * A new version of the *dev* (built on the corresponding `pyviz-dev` repository `gh-pages` branch) site should have been built. You should spot-check it, trying to find errors that could have occurred while the notebooks ran, for instance.
```

1. Announce the release (e.g., on Discourse, Gitter, Twitter).
1. If the release is important (e.g., not a bug fix release), it may be worth a blog post.

```{note}
Bumping the version of a package depends on the package's nature:
* Non-pure Python packages, e.g., Panel or GeoViews as they include JavaScript/TypeScript extensions, need their `package.json` and `package-lock.json` files to be updated, manually bumping the version in those files (e.g., from `1.9.5` to `1.9.6-a.1`). Note that the version scheme isn't the same as the Python version scheme.
* Pure Python packages (and non-pure Python packages once manually bumped) just need a new tag, done with `git tag -m "Version 1.9.6 alpha1" v1.9.6a1 master`.
```

```{note}
Development releases are cheap, don't hesitate to make as many as required! It's also fine if a development release only makes it to one of the two distribution platforms (PyPi or Anaconda.org). They're not meant to be for end-users but development purposes only.
```

```{tip}
You can tag a release from any branch, not necessarily from the main one. This is useful if you have to maintain multiple versions simultaneously (e.g., 2.* and 3.*).
```

```{tip}

> If you push a tag by mistake or the wrong tag and are lucky enough to spot that instantly, you should hurry up (really!) and cancel the *Build* and *Documentation* workflows before anything gets published/deployed. If you manage to do that, you can then remove the faulty tag.

Reviewer: how to cancel the workflows?

Author: Good point, adding a link to the relevant Github docs.

```suggestion
If you push a tag by mistake or the wrong tag and are lucky enough to spot that instantly, you should hurry up (really!) and [cancel](https://docs.github.com/en/actions/managing-workflow-runs/canceling-a-workflow) the *Build* and *Documentation* workflows before anything gets published/deployed. If you manage to do that, you can then remove the faulty tag.
```
